Added amclient module: can hide approved transfers.

jrwdunham opened the pull request.

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+        /location/UUID/browse/.
+        """
+        def dec(thing):
+            return base64.b64decode(getattr(thing, 'encode')('utf8'))
```
Why not `base64.b64decode(thing.encode('utf8'))` ?

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+    def search_package(self, params):
+        """SS GET /search/package?<GET_PARAMS>."""
+        return self._call_url_json(
+            '{}/api/v2/search/package/'.format(self.ss_url), params)
```
Resources in the storage service already allow you to filter by certain fields. Can this be done using those instead of requiring the new /search/ endpoint? eg /api/v2/file/?package_type=AIP

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+    def aip2dips(self):
+        """Get all DIPS created from AIP with UUID ``self.aip_uuid``."""
+        dips = self.get_all_dips()
+        return [d for d in dips if
```
Can this be done by only getting DIPs that related to an AIP UUID using tastypie's filters? https://github.com/artefactual/archivematica-storage-service/blob/stable/0.9.x/storage_service/locations/api/resources.py#L434 should be relevant.

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+        url = '{}/api/v2/file/{}/download/'.format(self.ss_url, self.dip_uuid)
+        r = requests.get(
+            url,
+            params={'username': self.ss_user_name,
```
Should this use `_ss_auth`?

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+        else:
+            print('Unable to download DIP {}'.format(self.dip_uuid))
+
+    def reingest(self):
```
These functions look like a duplicate of #17?

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+                  ' {}.'.format(self.aip_uuid))
+        self.approve_reingest(self.aip_uuid)
+
+    approve_reingest_retry_count = 5
```
This constant should be inside the function that uses it, or at the top of the class with the rest of the constants. Since it's only used once I'm inclined to say the former.

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+        time.sleep(5)
+        # List available ingests
+        get_url = '{}/api/ingest/waiting'.format(self.am_url)
+        params = {'username': self.am_user_name, 'api_key': self.am_api_key}
```
Should this use `_am_auth`?

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+        # List available ingests
+        get_url = '{}/api/ingest/waiting'.format(self.am_url)
+        params = {'username': self.am_user_name, 'api_key': self.am_api_key}
+        waiting_transfers = requests.get(get_url, params).json()
```
Should this use `_call_url_json`?

jrwdunham commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+            raise TypeError("expect bytes or str, not %s" %
+                            type(filename).__name__)
+
+    def fsdecode(filename):
```
Ok. Removed.

jrwdunham commented on transfers/amclient.py:
```diff
@@ -0,0 +1,518 @@
+
+    def hide_transfer(self, transfer_uuid):
+        """GET transfer/<UUID>/delete/."""
+        return requests.delete(
```
Ok. I've modified call_url_json to a func (no need to be a method) and it takes a method string param and calls requests.request instead of requests.get and it logs the method.

Hwesta replied:
Looks good!

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,487 @@
+KNOWN BUG: if you call ./amclient.py with no args/options, you'll get an error.
+See https://github.com/aliles/begins/pull/60 for the PR that fixes this on Py3.
+"""
+from __future__ import print_function, unicode_literals
```
:100:

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,487 @@
+    if method.upper() in ('GET', 'POST', 'DELETE', 'PUT', 'HEAD', 'OPTIONS'):
+        method = method.upper()
+    else:
+        method = 'GET'
```
Could this if/else be `method = method.upper()`?  That would allow eg PATCH, requests already defaults to GET, and so is the default parameter of this function.

jrwdunham replied:
Sounds good.

Hwesta commented on transfers/amclient.py:
```diff
@@ -0,0 +1,487 @@
+        result['directories'] = [dec(d) for d in result['directories']]
+        result['entries'] = [dec(e) for e in result['entries']]
+        result['properties'] = dict((dec(k), v) for k, v in
+                                    result['properties'].items())
```
Dictionary comprehensions are pretty great. This can be `{(dec(k): v for k, v in result['properties'].items()}`

jrwdunham replied:
Hereby consider me officially a dict comp convert.

Hwesta commented on tests/test_amclient.py:
```diff
@@ -0,0 +1,344 @@
+
+
+def is_uuid(thing):
+    return UUID_PATT.search(thing) is not None
```
This is a duplicate of what's in amclient - should probably just use that one.

jrwdunham commented on tests/test_amclient.py:
```diff
@@ -0,0 +1,344 @@
+    usage.
+    """
+
+    def __enter__(self):
```
ok.

jrwdunham commented:
Ok. I think I've addressed this round of concerns. I've changed the CLI to use argparse again. I got subcommands working. What do you think now?

Do you think tests targeting the command line interface are necessary or overkill?

Hwesta commented:
Nice job on the subcommands! 

> Do you think tests targeting the command line interface are necessary or overkill?

Probably overkill.

In python3, json.dumps() fails when the dict contains bytestrings (eg. in the command `transferables`).  "Works" in python2 until it encounters non-ASCII.  The results should be encoded (utf8?) before printing.  Interestingly this works for `unapproved-transfers` even when unicode provided, and none of the other commands seem to have this problem because the filenames are sanitized.

jrwdunham commented:
Nice catch about the json.dumps fail. I believe that simply modifying the return of ``b64decode_ts_location_browse`` to ``return thing.decode('utf8')`` in solves the issue. The ``return response.json()`` in ``_call_url_json`` should return Unicode (<str>) in all other cases.

Hwesta commented:
Looks good!  Breaks on non-unicode (& non ASCII) encodings, but that's not really a concern right now.
